### ISSUE TITLE
Reorganising LMMS, renaming editors and classes and everything in between (cont. #120)

### ISSUE THREAD

**Issue by ryuukumar** [LMMS/lmms#5592]
**If you are here because your PR is mentioned** then check the instructions at the bottom of the comment.

This is an issue continued on from #120. We're mostly following on from what was decided there, though if you have something to say, feel free to.

These are the draft steps that we will have done one-by-one in order to achieve this massive task:

- [x] Freeze code on December 5th, 2020
- [x] Clear PR backlog by January 25th, 2021
  - [x] Ping devs of older PRs via the table in this issue
  - [x] Review and test PRs to be merged before reorganization
    - [x] #5868 MIDI Range
    - [x] #5522 Microtuner
    - [x] #5657 Play automation pattern with controller
    - [x] #5846 Carla Automation
    - [x] #5826 Improved Oscillators
    - [x] #5735 Local paths and project bundles
    - [x] #5873 Update flanger effect
    - [x] #5923 Automation fine tuning with double click
- [x] Format code...
  - [x] with format PRs (#5791, _[request to not push any more such PRs](https://github.com/LMMS/lmms/issues/5592#issuecomment-740187644)_)
  - [x] with dos2unix (#6198, must be done prior to clang-format)
  - [x] with "include-what-you-use" (#5860)
  - [x] with Clang + CI (#4690, [comment](https://github.com/LMMS/lmms/issues/5592#issuecomment-738513836), latest approach at #6323) - It has been agreed on not doing an automatic format, but leave the decision to the authors of individual PRs
  - [x] However, we do agreed on a few special clang-tidy fixes:
    - [x] modernize-redundant-void-arg (https://clang.llvm.org/extra/clang-tidy/checks/modernize/redundant-void-arg.html)
    - [x] modernize-use-bool-literals (https://clang.llvm.org/extra/clang-tidy/checks/modernize/use-bool-literals.html)
    - [x] modernize-use-emplace (https://clang.llvm.org/extra/clang-tidy/checks/modernize/use-emplace.html)
    - [x] modernize-use-equals-default (#6450) (https://clang.llvm.org/extra/clang-tidy/checks/modernize/use-equals-default.html)
    - [x] ~~modernize-use-equals-delete (https://clang.llvm.org/extra/clang-tidy/checks/modernize/use-equals-delete.html)~~ no occurrences in our code
    - [x] after modernize-use-equals-default has run, follow it up with performance-trivially-destructible (https://clang.llvm.org/extra/clang-tidy/checks/performance/trivially-destructible.html)
    - [x] Update `.clang-tidy` (#6449)
  - [x] Get rid of the most clazy warnings. Currently, only #6445 .
- [x] Rename PRs
  - [x] Rename `MidiTime` -> `TimePos` (#5684)
  - [x] 'Clip' rename
    - [x] Rename `TCO` -> `Clip` (#6226)
    - [x] Rename `Pattern` -> `MidiClip` (#6226 + PR for data file tag)
    - [x] Rename `*TCO` and `*TCOView` -> `*Clip` and `*ClipView`  [* stands for any suffix eg. BB, Automation] (#6226)
    - [x] Rename `BB*` -> `Pattern*` (#6284)
    - [x] Adjust `DataFile` (#6309)
  - [x] Mixer rename
    - [x] **Rename `Mixer` -> ?** (Discussion at #6089)
    - [x] Rename `FX-Mixer` -> `Mixer`
  - [x] Rename any other remaining editors
- [x] Other major changes
  - [x] Split track classes into one file per class 
    - [x] `Track.cpp` (#5806)
    - [x] `AutomationTrack.cpp` (#5983)
    - [x] `BBTrack.cpp` (#5985)
    - [x] `InstrumentTrack.cpp` (Wait for #5522)
    - [x] `Pattern.cpp` (#5986)
    - [x] `SampleTrack.cpp` (#5987)
  - [x] Reorganise all files into better places & folders (_requires some discussion_ #5944)
    - [x] Rename `LedCheckbox.h` to `LedCheckBox.h` (the class is named `LedCheckBox`), and using a common file naming scheme for `stereo_enhancer...` vs `stereoenhancer...` (same for `stereo_matrix`)
    - [x] ~~Move Mixer, Controller Rack, and Project Notes windows to the `gui/editors` folder~~ removed since this folder was already well defined (Editor.cpp and subclasses)
    - [x] Move dialogs (that open in an external window) into `gui/dialog`. Delete the now-empty `gui/forms` folder
    - [x] Organize `widgets` folder (Possibly covered in #6374)
      - [x] Folder for `instrument` classes
      - [x] Folder for `track` classes
      - [x] Organize remaining classes by [reusability](https://github.com/LMMS/lmms/discussions/5944#discussioncomment-1371450) or [area of use](https://github.com/LMMS/lmms/discussions/5944#discussioncomment-2398444)?
  - [x] Add a namespace for LMMS (PR #6174 is ready to review, discussion at #6086, [an example](https://github.com/LMMS/zynaddsubfx/pull/16))
- [x] Reopen code for fresh PRs (or rebased old ones)

Feel free to drop suggestions, things you would want to see, etc. here. If you have something that would require a major changed in code like this renaming project (like maybe moving to C++17?) then please do tell us, because I think it's better to finish all major changes in one go rather than have a bunch of code freezes for each change.

----

### The section where I mentioned your PR

If your PR was mentioned by this issue, then ping back on your PR and mention this issue so I can change the state from "Waiting for Response" to whatever it really is. I'll probably ping all PRs which haven't responded.

Edit: Considering that new PRs are being added and merged, I'm changing this list to be a tracker of older PRs so that they don't get lost in pages and pages of PRs.

I've not tagged some PR authors because it was their first PR, and the tag system didn't allow me to tag them yet. These PRs are almost all stale and inactive for at least a year.

<details>
<summary>Expand table of PRs</summary>

| PR | Last Active | Has conflicts? | Stale/Active/Ready |
|---|---|---|---|
| #2068 | - | - | Waiting for Response from @michaelgregorius  |
| #2505 | - | - | Waiting for Response |
| #2944 | - | - | Waiting for Response from @BaraMGB |
| #3115 | - | - | Waiting for Response from @liushuyu |
| #3117 | - | - | Waiting for Response |
| #3134 | - | - | Waiting for Response from @liushuyu |
| #3224 | - | - | Waiting for Response from @tonychee7000 |
| #3467 | - | - | Waiting for Response from @liushuyu |
| #3532 | 2019 | No | (Feature Merge) Stale |
| #4027 | - | - | Waiting for Response from @tresf |
| #4067 | - | - | Waiting for Response from @tresf |
| #4232 | 2019 | Yes | (Feature Merge) Stale |
| #4357 | - | - | Waiting for Response |
| #4367 | - | - | Waiting for Response from @SecondFlight |
| #4369 | - | - | Waiting for Response from @SecondFlight |
| #4397 | 2020 | Yes | In Progress/**Ready** |
| #4412 | - | - | Waiting for Response |
| #4441 | - | - | Waiting for Response from @BaraMGB |
| #4443 | 2020 | Yes | (Feature Merge) In Progress |
| #4571 | - | - | Waiting for Response |
| #4635 | - | - | Waiting for Response from @DouglasDGI |
| #4662 | 2020 | Yes | Stale/In Progress |
| #4674 | 2018 | No | (Feature Merge) Stale/**Ready** |
| #4690 | - | - | Waiting for Response from @JohannesLorenz |
| #4894 | 2020 | No | **Ready** (needs review) |
| #4994 | - | - | Waiting for Response from @Reflexe |
| #5000 | - | - | Waiting for Response from @DouglasDGI |
| #5098 | - | - | Waiting for Response from @michaelgregorius |
| #5147 | 2020 | No | In Progress |
| #5158 | - | - | Waiting for Response from @artur-twardowski |
| #5161 | - | - | Waiting for Response from @DouglasDGI |
| #5166 | - | - | Waiting for Response from @iansannar |
| #5168 | - | - | Waiting for Response from @DouglasDGI |
| #5174 | 2020 | Yes | In Progress |
| #5197 | 2019 | No | Stale |
| #5202 | 2020 | Yes | **Ready** (needs review) |
| #5204 | 2020 | Yes | Stale/**Ready** (needs discussion) |
| #5229 | 2020 | No | In Progress (needs testing) |
| #5230 | 2019 | No | **Ready** (waiting for #5229) |
| #5236 | 2020 | No | Stale/**Ready** |
| #5239 | 2019 | Yes | Stale |
| #5247 | - | - | Waiting for Response from @Reflexe |
| #5261 | 2020 | Yes | Stale (needs review) |
| #5274 | - | - | Waiting for Response from @fschuelke |
| #5291 | 2020 | No | Dropped |
| #5292 | 2020 | No | **Ready** (needs review & testing) |
| #5356 | - | - | Waiting for Response from @artur-twardowski |
| #5357 | - | - | Waiting for Response from @komodor1 |
| #5436 | - | - | Waiting for Response from @Reflexe |
| #5458 | - | - | Waiting for Response from @DouglasDGI |
| #5500 | - | - | Waiting for Response from @artur-twardowski |
| #5502 | - | - | Waiting for Response from @thmueller64 |
| #5503 | 2020 | No | In Progress (waiting for review) |
| #5510 | - | - | Waiting for Response from @tresf |
| #5516 | - | - | Waiting for Response from @EmoonX |
| #5522 | 2020 | Yes | **Ready** (needs review and testing) |
| #5524 | 2020 | No | **Ready** (needs review) |
| #5544 | 2020 | No | Stale/In Progress |
</details>

**Comment by qnebra** [comment]
Best way to know that PR is active or not is just to peek into commit history. If there is nothing happen in more than 1 year then PR is stale.

**Comment by russiankumar** [comment]
When there's 87 PRs... it would be a lot easier to have devs help too. Don't wanna be doing the whole thing myself xD

**Comment by claell** [comment]
Hm, I know that #5261 is probably not stale, however I don't know whether the dev currently has time. The problem with some PRs is that it takes much time until the code is reviewed and when that finally happens the dev is no longer free to address the requested changes. Two weeks might be a pretty tough goal, but let's see.

Can we use some automation to see which PRs will be affected and maybe also some automation for the refactoring that can be easily applied to outstanding PRs? I think about using something like documented commands for a Unix terminal with regular expressions for everything. That way, the refactoring is both easier to understand, to adjust and reproducible. It then could be applied to open PRs pretty easily.

**Comment by russiankumar** [comment]
> which PRs will be affected

I'm pretty sure every PR will be affected considering the size of these changes

> documented commands for a Unix terminal with regular expressions for everything.

Yes, that has to be what will be used (like maybe a find+replace across all files) because doing everything manually will take forever. If you have suggestions on how to go about this, would love to hear them.

**Comment by Spekular** [comment]
With regards to my PRs:
- #5117 isn't stale, I intend to get it merged before this reorganization
- #5427 is a draft so it can stay marked as stale
- #5524 is in limbo. Its pretty much in the same state as #5117, it's just less important to me to get merged

**Comment by russiankumar** [comment]
Updated, thanks!

**Comment by Spekular** [comment]
In regards to this
> If you have something that would require a major changed in code

I say this is a good opportunity to split the Track classes (including subclasses like SampleTrack, AutomationTrack) into separate files. Those files are *huge* right now and among other things it makes it harder to keep track (pun) of which members come from which classes. Potentially making compilation more granular is a big plus too.

**Comment by russiankumar** [comment]
Yeah, it definitely is HUGE. How about three files (+ includes): `Track` (Track & TrackView), `TrackWidget` (TrackOperationsWidget, the likes) and `Clip` (current TCO and TCOView -> Clip & ClipView)

**Comment by Spekular** [comment]
I'm a fan of one file per class, but if that's not a popular proposal then three would at least be a step in the right direction.

**Comment by claell** [comment]
> Yes, that has to be what will be used (like maybe a find+replace across all files) because doing everything manually will take forever. If you have suggestions on how to go about this, would love to hear them.

Maybe we can come up with one or several bash scripts that contain the relevant commands.

**Comment by russiankumar** [comment]
Do approvals from people without write-access count towards mergeability? Github says not but maybe the mods here take it differently?

**Comment by IanCaio** [comment]
My PRs:

- #5503 : Stale. It's functional and there are not any requested changes, but I think it didn't raise much interest in people. Wouldn't mind if it didn't get merged.
- #5547 : Not stale. Functional and no current requested changes unresolved. Just waiting for review to get merged (it's a small PR so it shouldn't be a hard review).
- #5581 : Not stale. Functional and no current requested changes unresolved. Relatively recent PR, so still under testing and review.

I think the idea of renaming the editors and maybe reorganizing some things is good, but I want to express a small concern:

There are not that many active devs with merge priviledges and the amount of PRs makes their job difficult. Specially if we balance that there are PRs that are well written, making them easier to review, but others can be a bit of a mess and take way longer (or even might not be mergeable). Also, some PRs are as small as _< 50_ lines, others have _> 1000_ lines of code. Trying to shorten the review process can be dangerous in the sense that some bugs or badly written code can be merged, and then we will have a big mess to deal with later. Simply allowing people with no write-access approvals count to merge PRs is also dangerous in that sense, because we can't be sure everyone knows the coding guidelines and code base well enough to make a good decision, or even how committed they are in making a good review. That being said, I think one week is a __way__ too optimistic timeline for finishing the PR reviews.

As for the process of doing the refactor itself, I see two ways of it being done:

- Someone can take the burden and work on the refactor alone, but frequently be updating the PR with the changes and waiting for reviews from those particular updates before continuing. That would make the reviewing process easier, because mistakes would be corrected from the beginning before they grow more complex. It's also easier to review a couple hundred lines weekly than thousands of lines in a month.
- We can see how many devs/contributors volunteer to the cause, discuss in very fine details the changes that will be made and assign files for each volunteer with a timeline (a light one that takes into account that most of us have jobs and life appointments as well). At the end of that timeline, we all submit those changes so they can be reviewed for a while. New files are assigned and so it goes.

I particularly prefer the second option, this doesn't sound like a one man job.

I don't know if I'm sounding too precautions, this refactor is a really good thing. I just don't want to see it cost the current stability we have on LMMS.

**Comment by Veratil** [comment]
My PRs:
* #5123 approved but needs testing, no conflicts
* #5124 approved but needs testing, no conflicts
* #5174 need to update IRT one review, has conflicts easy to resolve
* #5196 needs to be reviewed and tested, has conflicts easy to resolve
* #5204 Proof of concept, meant for discussion, but works (at the opening of the request)
* #5229 Step 2 in removing global automation from #403, needs a lot of testing, no conflicts
* #5230 Step 3 in removing global automation from #403, needs testing after #5229 is merged, no conflicts
* #5253 we should probably merge this before I keep adding to it 😂, no conflicts

**Comment by he29-net** [comment]
>Ping all authors of all current PRs asking them to wind up within two weeks since now (done by the table)

@russiankumar Just FYI, this "ping" did absolutely nothing in my case. I happen to be obsessively reading the GitHub log on Discord, so I noticed this issue, but I imagine most of the PR authors have better things to do and won't notice anything since it did not generate an e-mail notification.

So by marking all PRs that don't get any response as "stale", we may be simply discarding a lot of good and useful code from authors that are no longer active and won't notice a mention appeared on their PR -- if they have been waiting for test and review for months (or even years), I would assume they just gave up and forgot about it. Some of the PRs could be unfinished WIP, sure, but I wouldn't rely just on the fact that the author did not reply to a ping. Some of them may not even be able to, for whatever reason (lost the account, fell off a cliff, signed an NDA in their new job, ...).

The PR backlog is a _huge_ problem for LMMS in my opinion (in a way, there is already an issue talking about this: #4935) -- it slows down development if someone needs to wait for another PR to get through, it generates conflicts in newer PRs when something old finally gets merged, it demotivates new devs who have to wait a long time to see their contribution accepted, and it demotivates experienced devs who just see a big pile of work that won't stop growing.

So it's great that someone took the initiative, thanks! But I doubt it can be solved within a few weeks just "by the way" as a few bullet points in this issue. The backlog exists precisely because reviews take a lot of time and effort, and we have just a handful of experienced devs who can do it, all with their own lives, time budgets and motivations. Clearing the backlog will probably require quite some team effort and several months at least. That is, if we want to avoid the "mark everything as stale and flush it down the toilet" approach. Authors of the affected PRs would probably think twice before investing time into making another contribution that could end up the same...

The code freeze could be good idea (but only for new features, not bug fixes), it could cool things down enough to give the dev team time to work on the backlog. Although at the same time, it could discourage new potential devs, who would be unable to submit their contributions.

It's just a hard problem to crack either way.. :-/

---
As for PRs where I'm involved:
 - #5522 is my own and waiting for testing and review (but it seems to have a new merge conflict);
 - #4397 is from curlymorphic and waiting for merge of my branch to the PR and subsequent review of the changes (PhysSong already reviewed the original code and I already did some testing when fixing / finishing the PR).

**Comment by russiankumar** [comment]
Thanks for the PR updates, everyone.

Many of you have pointed out a concern that two weeks is nowhere near enough to finish reviewing and merging as many PRs as possible. I am convinced, so I suppose time till mid to end of August should be enough? That essentially does leave one month and a bit for everything.

As @he29-net pointed out, waiting for too less or too long are both problems. It's in our best interests to probably find a middle ground and finish merging most PRs. It's also enough time for all devs to respond (if they chance upon their PRs anytime soon). If they don't respond, then I can drop a message on their PRs to see whether they can finish it up or not. It's definitely bad if we ignore someone's hard work and just dump it. Since most probably these PRs will not be able to be merged after these changes, now has to be the time we do it.

It's probably going to involve a lot of work, and given the fact that write-access devs are few and usually busy, won't be an easy thing. But then again, it's best to solve the issue now before it expands even more into a big mess, discouraging new devs from developing.

I hope this shall leave enough time for most devs to ping back, and for as many (reviewed and tested) PRs to merge.

**Comment by softrabbit** [comment]
I noticed this issue by luck, really. It's not like I keep going back to read my PRs every now and then, I've always expected relevant activity to throw a mail in my direction.

Anyway, I'm guilty of these:
#5236: Stale, I guess. It took 5 months, but a better way to do it appeared. Which kinda points at the long process being a good thing?
#5239: Stale. Something I totally forgot about and never got around to finishing. Doesn't seem like a big thing to do properly after the changes.

**Comment by claell** [comment]
Since there is no email notification sent out the current way, maybe you can ping the devs by adding them to the first comment on this issue @russiankumar, like I tagged you in this comment. Hopefully that should trigger an email notification so the relevant devs are aware of this.

**Comment by russiankumar** [comment]
@claell sure, for those PRs which haven't responded I'll do as you say.

@softrabbit, could you clarify on whether you will finish the PRs or not? I'm marking them as stale in case you do want to restart work on them and get them ready for merge.

**Comment by tresf** [comment]
> Ping all authors of all current PRs asking them to wind up by mid/end August

Let's not do this.  Just focus on the large, active ones.   I need to turn it into a circus

**Comment by russiankumar** [comment]
Okay, as you say.

**Comment by tresf** [comment]
@DomClark made a good point on Discord about the formatting PR, that we should pair that with a CI style enforcement-type job to keep the style guidelines from drifting.  Might be an important precursor to large refactoring.

**Comment by russiankumar** [comment]
I think it's better to do that after we merged all the PRs that are going to be merged, then we don't need to redo it each time a PR is merged.

**Comment by tresf** [comment]
Yes as part of or as compliment of the code formatting PR.

**Comment by Cyp** [comment]
Didn't know #5304 hadn't been merged. Just fixed some merge conflicts.

**Comment by thmueller64** [comment]
#5502: In progress, will be finished this weekend.
#5549: Ready from my side, the PR fixes the issue. I'll try to incorporate the suggestions by @Veratil (gradients instead of colors) in time, but those aren't strictly necessary for fixing the issue.

**Comment by RebeccaDeField** [comment]
#5588 is going to be merged within the week, which will, in turn, make #4255 fine to close out (was going to after mine is merged in but if we want to speed things along I guess we can do so now?).

**Comment by he29-net** [comment]
@russiankumar I just noticed your table says that 4397 waits for 5522 -- I never said anything like that, these are completely unrelated PRs. o_O

4397 is waiting for someone with write access, who can merge my branch (where I finished and fixed the PR) into the branch of the original author (who can't work on it anymore).

**Comment by Spekular** [comment]
#5131 might be a candidate to include in the refactor.

**Comment by claell** [comment]
> > Ping all authors of all current PRs asking them to wind up by mid/end August
> 
> Let's not do this. Just focus on the large, active ones. No need to turn it into a circus

At least pinging the authors is only fair. Maybe they don't need to wind up their PRs until then, but not notifying them appropriately is not nice imho, especially given the fact that the reason many PRs are still open is partly on LMMS side (no offense).

In addition to notifying, I think it would be nice to make incorporating the made changes to outstanding PRs as easy as possible (as suggested earlier for example by having a bash script that contains all commands needed for the refactor and only has to be applied to the codebase).

**Comment by tresf** [comment]
> but not notifying them appropriately is not nice imho, especially given the fact that the reason many PRs are still open is partly on LMMS side (no offense).

My point was [already made in a few words](https://github.com/LMMS/lmms/issues/5592#issuecomment-662803238).   If PRs are waiting excessive times on reviewers, that's a valid concern, but I'm not sure it's related to pinpointing "large, active ones".  80/20 rule should apply.  If that's "not nice", it's not personal, it's just part of large changes.  Perhaps to put it more bluntly, no large change ever happens without making someone uncomfortable or upset.

That said, if anyone has PRs that should have been merged a long time ago, it's probably best to start a discussion on #devtalk on Discord.  We might be able to promote a few contributors to help review and merge assuming the current devs are onboard with it.

**Comment by tresf** [comment]
> having a bash script that contains all commands needed for the refactor and only has to be applied to the codebase

Note, if there's real conflicts, this is [rarely a one-liner](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/resolving-a-merge-conflict-using-the-command-line) but devs are always happy to help over on #techtalk. <3

**Comment by claell** [comment]
Alright, maybe I got the few words wrong and misunderstood your goal. I guess we agree on notifying the devs in general, then and your point was rather about whether it makes sense to try everything to merge before the deadline, even where the effort might not be worth the benefit.

Regarding the bash script: The idea was to have a script that should apply the refactoring to the branch of a PR and thus resulting in alot less merge conflicts that would need manual tackling.

**Comment by tresf** [comment]
> The idea was to have a script that should apply the refactoring to the branch of a PR and thus resulting in alot less merge conflicts that would need manual tackling.

GitHub has a visual merge conflict resolution that helps, but in my experience, if it's in conflict with GitHub it'll be in conflict on the command line and there's generally no series of small, short commands that fixes it (but rather the painstaking process of hand-editing the conflicts).   If I'm wrong, please share.

> whether [or not] it makes sense to try everything to merge before the deadline, even where the effort might not be worth the benefit.

Well, as an example, I was tagged for #5510 which is just a few lines and it's something that -- according to the reviewer -- is feature incomplete, so won't get revisited for a while.  Rebasing it is isn't a problem at all.  I can't speak on behalf of all of them though.

**Comment by russiankumar** [comment]
@claell actually considering the fact that we're doing more than renaming, we're actually rearranging files in new directories and splitting large files into more files and stuff, I really doubt whether it's possible to have a bash script do that.

I will update the table with updates ASAP, sorry for the delay.

**Comment by russiankumar** [comment]
Considering that the number of PRs is steadily holding at 80-ish, I'm not very sure if we will be able to resolve them all soon (especially with new PRs still being made). I suppose that it would make sense to make a call on whether to postpone this or entirely focus on closing existing PRs (and actually do as decided), else this change will never happen.

**Comment by JohannesLorenz** [comment]
Dom and me just finished #5517, feel free to mark it as solved/remove it from list.

**Comment by Veratil** [comment]
Merged #5253 (finally!)

**Comment by russiankumar** [comment]
Great, thanks for the ping!

**Comment by claell** [comment]
> If I'm wrong, please share.

We seem to talk about different things here. I was proposing a script that could apply the changes to PRs, so that no merge conflicts (or at least less) would occur.

I agree that there are probably several open PRs that are not ready to merge in the current state and with the author having no time to complete it currently (plus maybe also not so hard to rebase). I also share the opinion that those requests can be skipped then. I only was suggesting to give the authors a chance to get it in by making sure they are aware of this approach and assisting if they want to get it merged beforehand.

> considering the fact that we're doing more than renaming, we're actually rearranging files in new directories and splitting large files into more files and stuff, I really doubt whether it's possible to have a bash script do that.

I think it is possible also when it comes to rearrangement of files etc. The question is whether those scripts will also work with all open PRs since their codebase is different. At least for the renaming things I suggest to use a script, though. It still might be useful to also do the rest with a script to ensure reproducible and defined results.

**Comment by Veratil** [comment]
Merged #5123

**Comment by JohannesLorenz** [comment]
For many of the above PRs, "Waiting for response" is not too helpful. Most of the users here won't be aware of their own PRs until their names are listed there. For example, did @DouglasDGI realize that he still has (the long awaited) #5000 open (it's really a pain that this PR never got to master...)?

**Comment by JohannesLorenz** [comment]
Some devs asked if we should get a list of really important PRs. From the PRs that I know:
* #5619 and #5536 should be fixed before reorganizing, because they're both large (and both probably easy to finish)
* #5532 can be finished right after #5536 is being finished
* #4894 maybe
* #4662 - I might need to rebase the PR so (almost) only new files are changed
* #4635 - I need to help @DouglasDGI , maybe @Spekular can help, too
* #4443 - Maybe (not sure what's missing here, but it has 86 files and is named "Refactoring ...")
* #4027 - I have to finish this. It's open for so long. Shame on me!

Two [other PRs with "refactor" in it](https://github.com/lmms/lmms/pulls?q=is%3Apr+is%3Aopen+refactor) are candidates:
* #5627
* #5573

**Comment by russiankumar** [comment]
> Most of the users here won't be aware of their own PRs until their names are listed there.

I'd decided not to tag all the devs by the [request](https://github.com/LMMS/lmms/issues/5592#issuecomment-662803238) of @tresf. I can do that if required, though.

> Two other PRs with "refactor" in it are candidates

I'm sorry, I think I mentioned 'refactor' in the title in the wrong context... Nevertheless I still think it is an important PR.

**Comment by IanCaio** [comment]
As per requested on Discord, here are my other PRs besides the one listed on Johannes previous comment:

PRs that refactor code:

* #5660 Improvement to the upgrade routines

PRs that add new features:

* #5581 Adds support for MIDI CC events inside LMMS

The feature PR is one that I consider very interesting for LMMS, because it allows automating some plugins that are currently not automatable (at least not without some really complex workarounds). But recently a bug was found, which I might need some help figuring out, since it is probably related with other areas of the code. Anyways, even though a new deadline wasn't set yet I believe it should be fixed by then. The refactor one is finished, but requires testing.

There's another feature PR open but I didn't bother listing it because it adds a feature that is not consistent in all operating systems.

**Comment by tresf** [comment]
> macOS LV2: #5532 can be finished right after #5536 is being finished

I don't see how it's related to this PR. It's two one-liners, which should be easily mergeable after this is merged.

> #4027 - I have to finish this. It's open for so long. Shame on me!

Well, it's a lot of trial and error and you're doing a lot :D.  I'm also not sure it's warranted prior to this PR since it really only changes a few files (deletes a directory and adds some new upgrade routines).  If it's postponed, I don't think it will make the project as a whole suffer.  I think it's a trivial PR to rebase if done manually (resetting to HEAD and hand-committing the few minor changes).

**Comment by DomClark** [comment]
Would this be a good opportunity to introduce a namespace for LMMS, as mentioned in https://github.com/LMMS/zynaddsubfx/pull/16?

**Comment by JohannesLorenz** [comment]
> > macOS LV2: #5532 can be finished right after #5536 is being finished
> 
> I don't see how it's related to this PR. It's two one-liners, which should be easily mergeable after this is merged.

Agreed, but I put this on the TODO list to not forget #5532. It can be done immediatelly after #5536, so having it here won't stop us.

> > #4027 - I have to finish this. It's open for so long. Shame on me!
> 
> I think it's a trivial PR to rebase if done manually.

Agreed. I see it now after I cleaned up this branch. I removed it from the above list.

**Comment by claell** [comment]
Just saw #4893 mentioned. Unfortunately I don't have the time needed to invest into that in the foreseeable future. Feel free to close, takeover or leave stale :)

**Comment by russiankumar** [comment]
@claell considering that renaming FxMixer to Mixer is part of this issue's agenda, and that you have already had some success with it, I'd vouch for keeping it open but stale, so that we have a headstart later on. If anyone is against this idea, you could close the PR.

**Comment by Spekular** [comment]
Just noticed the "Create a new branch for all changes" point and I'd like to argue against it. If master is truly frozen there's no benefit to a separate branch. If master isn't truly frozen then working on a separate branch will just delay merge conflicts into the future, worsening them.

**Comment by russiankumar** [comment]
Sure! I had left it there with a request for discussing it, so if anyone else has anything to say, please do. I myself wouldn't mind either method.

**Comment by IanCaio** [comment]
The current deadline we have on clearing the PR log was December, but we didn't do anything to prevent new PRs to be submitted. Not sure Github allows closing the PR section temporarily, but if not maybe an announcement that PRs from date X forward will not be merged before the refactor and will need to be fixed later (could even create a label for it, like `after-refactor`, bug fixes being an exception).

I'm also guilty of submitting a couple difficult PRs to review recently and I'll try to compensate for that by focusing on reviewing from now on instead of writing more stuff :sweat_smile: . But what I mean is that if we are decided towards making this refactor happen soon maybe we will have to take some more active actions to set it in motion.

What I propose is:
- Every dev active on Discord or Github informs us if there's some pending PR they were already working on or something that they wish having merged before the feature-freeze that **is not already** in the PR log.
- We add it to a list and start the feature-freeze from that point on. Nothing besides the active PRs and those mentioned will be merged before the refactor.
- Set up a new deadline, since I think December might be a bit complicated at this point, focus on reviewing and merging the active PRs and the ones on the list we made earlier. It doesn't even need to be a short deadline, as long as the feature-freeze is happening we would always be getting closer to our goal. The importance of making an announcement (or ideally locking the PR section) is that we'd have to be a bit strict about the new PRs not being included.
- Set the refactor in motion.

I'm a newcomer on collaborating with a bigger project, so I'm not sure the best way to proceed with the division of tasks during the refactor itself. But I'd probably suggest we initially divide the work among the participating devs and have a week of coding followed by one week of peer-review for everyone's weekly progress. Everyone would work on their own pace, but we would be adding the code in the trunk style proposed by @Spekular on the end of the review week and resume work from there.

Since each change is differently spread in the code base, we'd discuss of an appropriate way to do each incrementally before starting work.

**Comment by JohannesLorenz** [comment]
I would propose January 10 as a deadline, as many coders probably have free days around the new year and this year, probably no one is going on vacation anyways. If we are done earlier, that's no problem either.

Also, I would vote for critical fixes to be allowed (e.g. fixes against sound that may kill your speakers). Such fixes are seldom anyways.

**Comment by tresf** [comment]
I think you guys are being too kind.  Let people rebase.  😈

**Comment by Spekular** [comment]
> Not sure Github allows closing the PR section temporarily

Doesn't look like it, but maybe something like [Repo Lockdown](https://github.com/marketplace/actions/repo-lockdown) would work if it can be configured to only reject new PRs. It'd be annoying to go through and re-open the valid ones though, so in the end maybe it's better to manually close invalid ones.

Perhaps we could (temporarily) update the README and some of the wiki pages to alert people about feature freeze? Maybe even the "feature request" issue template.

> Also, I would vote for critical fixes to be allowed (e.g. fixes against sound that may kill your speakers). Such fixes are seldom anyways.

Sounds reasonable to me as well.

**Comment by IanCaio** [comment]
> Also, I would vote for critical fixes to be allowed (e.g. fixes against sound that may kill your speakers). Such fixes are seldom anyways.

For sure, I meant to be broad when I said bug fixes, I'd consider that one! I had to replace my laptop speakers recently and I DO NOT want to have to go through that again :laughing: 

> I think you guys are being too kind. Let people rebase. smiling_imp

Just don't make me rebase #5712, I beg you! :rofl:  

> Doesn't look like it, but maybe something like [Repo Lockdown](https://github.com/marketplace/actions/repo-lockdown) would work if it can be configured to only reject new PRs. It'd be annoying to go through and re-open the valid ones though, so in the end maybe it's better to manually close invalid ones.
> 
> Perhaps we could (temporarily) update the README and some of the wiki pages to alert people about feature freeze? Maybe even the "feature request" issue template.

I see, that would probably be a bit of a hassle. We should make a very visible announcement then, on both discord and here on Github (wiki pages like "How to contribute" and maybe the root README.md) and create that `after-refactor` label so we can add them to the PRs that get submitted after the announcement anyways.

**Comment by tresf** [comment]
I strongly disagree with any proposal that attempts to manage a large refactor by blocking PRs.  Feature freezes are for merging, not for proposing change.

* Some believe the style change should go in first.
   * If that's the case, do it now.
* Some believe some larger PRs should be wrapped up before this is merged.
   * This was proposed in July.  It's too late, they'll have to be rebased.

Most large PRs are done by people that are perfectly capable of refactoring and if they're not a dev can take an hour to rebase the valuable ones when the OP won't.  This is what @Sawuare and I did over at https://github.com/LMMS/lmms/pull/4232.  It's tedious, but it's not difficult.

As far as I'm concerned the only people that get to hold this up are people that are very active in the PRs area and even then, the cutoff should be in days, not weeks.

If you really want to, we might be able to create a PR template that puts up a warning about this effort, but please, no more.  Don't discourage help.

**Comment by tresf** [comment]
>  I'd consider that one! I had to replace my laptop speakers recently and I DO NOT want to have to go through that again 😆

I feel this is an edge-case that shouldn't drive decisions in this PR.  If a speaker is damaged because of bad DSP code it can and will happen again, especially when it's 3rd party code.  From what I understand this was testing LV2 and one plugin went haywire (#5767).  I don't know how we can avoid this and it sucks it happened, but the project has money to help compensate you for your equipment loss.  Perhaps In the future we can have an fx filter on master which avoids speaker-damaging audio, but that seems like a separate topic.

**Comment by IanCaio** [comment]
Valid points, I didn't think about the side-effect of discouraging new collaborators with the PR lock suggestion. We can go through with the announcement + labeling idea, if everyone agrees with it, so at least we know what is to be merged before and after the refactor.
 
> I feel this is an edge-case that shouldn't drive decisions in this PR. If a speaker is damaged because of bad DSP code it can and will happen again, especially when it's 3rd party code. From what I understand this was testing LV2 and one plugin went haywire (#5767). I don't know how we can avoid this and it sucks it happened, but the project has money to help compensate you for your equipment loss. Perhaps In the future we can have an fx filter on master which avoids speaker-damaging audio, but that seems like a separate topic.

Oh, my blown speakers were completely unrelated to this issue! I'm really sorry for the misunderstanding, I was just joking about how hard it is to replace a laptop's speaker :flushed:

**Comment by M374LX** [comment]
#4894 is ready for review.

**Comment by M374LX** [comment]
Another file needing to be refactored is main.cpp. I also suggest moving it from ```src/core``` to ```src``` not only because this would give more visibility to the file in which the execution of the program begins, but also because the file inevitably depends on code from both ```core``` and ```gui```, even in an ideal world where ```core``` is fully independent from ```gui```.

**Comment by JohannesLorenz** [comment]
Sorry for the close, I was just tired and added `Fixes 5592` into the commit message of #5806 . Reopened.

**Comment by M374LX** [comment]
While there was previously a convention to leave four blank lines between each method implementation, the current conventions do not say anything about the number of blank lines.

I think one blank line between methods of the same class is enough, while we could use possibly two between different types of declarations on a file (like includes, forward declarations, and method implementations) and also between methods of different classes implemented in the same file. Example:
```C++
#include <QApplication>
#include <QWidget>

#include "foo.h"
#include "bar.h"


class Class1;
class Class2;


void Class1:method1()
{
}

void Class1:method2()
{
}


void Class2:method1()
{
}

void Class2:method2()
{
}
```

In the case of headers, I think one blank line between classes is enough:
```C++
class Class1
{
	void method1();
	void method2();
};

class Class2
{
	void method1();
	void method2();
};
```

**Comment by JohannesLorenz** [comment]
As there are new formatting PRs (whitespace fixes, variable renames ...) recently: I think we should stop writing new formatting PRs before we do the `clang-format` runs. It's only duplicate work and we want to avoid new PRs anyways.

**Comment by russiankumar** [comment]
Pinging for some things that are still open for discussion:

- Reorganise all files into better places & folders
  - What are the possible folders to be made?
  - Should we update the `include` folder? If so, how should it be organised?
- Clear out space for renamings
  - Are there any changes that need to be done in order for the mentioned renamings to be done?
- Moving to C++17
  - Is it worth it? What changes, if any, will have to be made to the codebase?

Feel free to add things to this comment if there's anything you need to be discussed.

**Comment by Veratil** [comment]
>       * What are the possible folders to be made?

This would be better figured out by going file by file and making tags of what purpose it falls under. At that point we can start deciding better folder structure.

>       * Should we update the `include` folder? If so, how should it be organised?

The same way we have the `src` tree IMO.

>     * Moving to C++17
>       
>       * Is it worth it? What changes, if any, will have to be made to the codebase?

We *could* do this, but I don't know what restrictions we'll have for distro support. For planning for Qt6, this is a must update though since it only supports C++17 and up.

**Comment by JohannesLorenz** [comment]
Just for fun, I made a C++17 PR: #5838. Almost all CI failed (but it works on my local computer).

I recently only upgraded LMMS to use C++14 (instead of 11), and IIRC it was very close to not working. So I don't expect that we can do C++17 in the next 3 years.

**Comment by JohannesLorenz** [comment]
Btw, what's the reason to rename `Pattern` to `Clip`? Has there been discussion about this?

**Comment by Veratil** [comment]
> 
> 
> Btw, what's the reason to rename `Pattern` to `Clip`? Has there been discussion about this?

IIRC, it's because `BB` should be renamed to `Pattern` since you can replicate the BB patterns and repeat them all over, but a normal pattern is its own unique "clip".

**Comment by Spekular** [comment]
> Btw, what's the reason to rename `Pattern` to `Clip`? Has there been discussion about this?

I think it's because we should rename TCO to Clip, and the "pattern" classes inhereit from TCO

- Clip is more standard (used in both Ableton and FL's manuals to refer to what we've called "TCO")
- It's shorter than "TrackContentObject", removing the need to use the abbreviation "TCO" everywhere. It's also easier to say than the full name or T-C-O, so I can stop reading it as "Teeko" :P

**Comment by Sawuare** [comment]
> Clip is more standard (used in both Ableton and FL's manuals to refer to what we've called "TCO")

Audacity similarly uses it.
https://manual.audacityteam.org/man/audacity_tracks_and_clips.html

**Comment by musikBear** [comment]
Someone rather aggravated asked me:

> "Why do you have *record buttons* and menu-options of *record* , if lmms **cant** record ?! (+angry smiley)"

That is actually a valid question. Most would understand the word *record* as recording of *sound*. Should the word record be changed in lmms' UI to *capture* or a similar word, and then we would have the word -**record** in the cold-store, ready for use, when lmms' get real sound record-features ?

**Comment by musikBear** [comment]
> Clip is more standard (used in both Ableton and FL's manuals to refer to what we've called "TCO")

Clip is (also) used in the meaning of a *looping*-sample. All samples in the most prominent *Looper-software* **SONY-asid**, are called *Clips*.
For *Pattern* I would favour **Block** because it *is* one, so the term makes sense in both language and visual context

**Comment by Spekular** [comment]
> > "Why do you have _record buttons_ and menu-options of _record_ , if lmms **cant** record ?! (+angry smiley)"

> That is actually a valid question.

Not really, considering it's based on the false premise that LMMS doesn't have any recording functionality, even though it does have MIDI recording.

> Most would understand the word _record_ as recording of _sound_. Should the word record be changed in lmms' UI to _capture_ or a similar word

Personally I prefer to understand words by via their actual definition and usage. Let's not use a less standard ([FL](https://www.image-line.com/fl-studio-learning/fl-studio-online-manual/html/recording_scores.htm#:~:text=Turn%20on%20the%20Note%20recording,Step%20Editing'%20should%20be%20off.), [Ableton](https://www.ableton.com/en/manual/recording-new-clips/)), clunkier word.

> then we would have the word -**record** in the cold-store, ready for use, when lmms' get real sound record-features ?

As mentioned above, we have real MIDI recording already. What we don't have is audio recording. Specifying what's being recorded is much clearer than using some arbitrary record/capture distinction.

**Comment by IanCaio** [comment]
I also disagree. It doesn't take much reasoning to realize that the "Record" button on the PianoRoll is to record MIDI and not raw audio. We can't start being so nit-picky about terminology. Think our (limited) efforts would be much better spent on more important things like new features, functionality and bug fixes.

As for the `Clip` name, I can't give much opinion because I started producing on LMMS, so I'm used to the TCO name. But I don't get confused by the word "Clip" and afaik that's an old discussion that took long enough for a consensus to be decided, so I don't think it's worth it to bring that on the table again :grimacing:

**Comment by cyber-bridge** [comment]
About renaming `Pattern` to `MidiClip`; Maybe a more generic name like `NotesClip` would be better?

**Comment by Spekular** [comment]
I've taken the liberty of adding the PR queue listed on Discord to the checklist here.

**Comment by SeleDreams** [comment]
Honestly I feel like Pattern is one that felt pretty straightforward to me

**Comment by Spekular** [comment]
I split `Split track classes into one file per class` into one item per file, for now. IMO we should split any other files that have multiple classes as well, but it doesn't look like that's been mentioned in this conversation. Splitting all the track files was though, and it looked misleading to me to mark that as "done" just because `Track.cpp` was split.

**Comment by Spekular** [comment]
I'll be splitting the track classes and merging when CI passes, one PR and commit per split. If anyone is opposed, let me know.

**Comment by musikBear** [comment]
> I'll be splitting the track classes and merging when CI passes, one PR and commit per split. If anyone is opposed, let me know.

Would it be a good idea, to make a top-comment, with something like:
"This class used to be combined with x,y,z,w ... but x,y,z,w is now split into the single classes  aaaa, bbbb, cccc .."
For each new, Just to keep a memo for reference ?

**Comment by Spekular** [comment]
> Would it be a good idea, to make a top-comment [...]

I don't think that should be necessary. It's quite clear that they're related from the names, imports, and so on. It's always possible to go back in the git history, too, so someone who wanted to check how they used to look could simply look at the code for 1.2.

**Comment by Veratil** [comment]
> 
> 
> > Would it be a good idea, to make a top-comment [...]
> 
> I don't think that should be necessary. It's quite clear that they're related from the names, imports, and so on. It's always possible to go back in the git history, too, so someone who wanted to check how they used to look could simply look at the code for 1.2.

Update the tasks with PR#s though? :)

**Comment by SeleDreams** [comment]
I just had something in mind, I feel like something that would make it much easier to implement new instrument, track types and all such as my UTAU integration would be if LMMS used more interfaces, right now LMMS uses mostly direct class references for most stuff, which requires quite a lot of inheritance and overrides, by having interfaces for notes, patterns and all it would make it easier to do dependency injection

**Comment by SeleDreams** [comment]
for instance if we take the piano roll as an example, vocal instruments would need to display lyrics and have them editable, rather than using ifs in the piano roll class to determine how to render notes, it could simply rely on a render function in the interface of the piano roll note and it would render with or without editable lyrics

**Comment by SeleDreams** [comment]
also when it comes to the stuff getting renamed, I'd suggest to rename detuning to pitch bend
it caused me lots of confusion when I started working on lmms because most other softwares I used called pitch bend the edition of the pitch of notes rather than detuning

**Comment by JohannesLorenz** [comment]
I would like to add #5860 to the reorg topic list. @russiankumar is that OK?

**Comment by JohannesLorenz** [comment]
At **all developers**, the following 2 reorg topics need to be **discussed** (**blockers**!). Please all help discussing, so we can start reorg:

* #5944 Source File Reorganization
* #5948 Clang tools testing

There are still 3 open PRs, but they will all not block:

* #5522 Microtuner (first review done, rework in progress)
* #5826 Improved Oscillators (first review done, rework in progress)
* #5846 Carla Automation (IMO and in @PhysSong 's opinion this might be merged without further review)

**Comment by JohannesLorenz** [comment]
> vocal instruments would need to display lyrics and have them editable, rather than using ifs in the piano roll class to determine how to render notes, it could simply rely on a render function in the interface of the piano roll note and it would render with or without editable lyrics

I'm in favour of such changes, as long as the virtual functions don't make runtime significantly worse. I just wonder if that kind of change will cause so many merge conflicts that it must be done during the reorg?

**Comment by russiankumar** [comment]
> I would like to add #5860 to the reorg topic list. @russiankumar is that OK?

@JohannesLorenz no issues, feel free to add anything relevant to this issue. My apologies for the delayed response, I am quite busy with personal matters right now.

**Comment by JohannesLorenz** [comment]
Devs, please also discuss #6086 and #6089 .

**Comment by JohannesLorenz** [comment]
There are a lot of side-issues linked here. In order to keep track, I created a project "[Reorg](https://github.com/LMMS/lmms/projects/13)" to put them all in. Devs, please feel free to subscribe to the issues linked.

Unfortunately, it seems like you cannot add discussions to a project, so click the "Discussions" tab (next to "Issues" and "PRs") to see them all.

**Comment by JohannesLorenz** [comment]
Like discussed in the discussion #6089 , we will rename `Mixer` to `AudioEngine`. Any volunteers?

**Comment by M374LX** [comment]
@JohannesLorenz 
I volunteer to rename Mixer to AudioEngine.

**Comment by tresf** [comment]
> * [ ] with dos2unix (must be done prior to clang-format)

FYI, GitHub allows CR/LF fixing through `.gitattributes`.  This will prevent stuff like `^M` from entering the codebase on clone and will make sure to strip any unwanted on push.  The downside is some old-timers using [older `notepad.exe` versions](https://stackoverflow.com/a/50241693/3196753) may suffer, but I don't see that as a concern in 2021 when Notepad++, VSCode, Atom, Sublime, QtCreator, etc all handle unix files on Windows just fine.

**Comment by JohannesLorenz** [comment]
> GitHub allows CR/LF fixing through `.gitattributes`

Sounds good. So we take it, but still have to run a tool like `dos2unix` over it before. Correct?

Btw, PR backlog is cleaned now. Next task here is probably splitting `InstrumentTrack.cpp` (as described above). It was waiting for #5522, which is now finished. However, I'd like to wait with that task, until we have merged the large #6127 PR.

**Comment by M374LX** [comment]
@JohannesLorenz Again, I volunteer.

**Comment by JohannesLorenz** [comment]
What's next?
@irrenhaus3 still has #6174 (namespace `lmms`) in progress, so I'll wait with clang-format for that.
Meanwhile, there are still a few rename PRs to be done (mentioned in the top post).

**Comment by M374LX** [comment]
In the next days, a plan to review #6044 and decide what to do.

**Comment by JohannesLorenz** [comment]
@M374LX Any plans on #6044? Can you estimate when you will have time for it?

**Comment by M374LX** [comment]
@JohannesLorenz Next saturday. :+1:

**Comment by JohannesLorenz** [comment]
The next things on our queue are:

1. Fix savefiles corresponding to #6226
2. Rename `BB*` -> `Pattern*` (had previously been blocked by #6226, became available now)
3. Possibly rename source files (though this might be delayed?) (#5944 , @Spekular has written a summary)

Who wants to pick one of these topics?

**Comment by allejok96** [comment]
🙋1?

**Comment by Spekular** [comment]
I've updated the main post with a summary of the files to move, but as pointed out by Johannes it may be best to wait with this until the other steps are done.

**Comment by allejok96** [comment]
Wait, so savefiles are already changed (and broken)?
https://github.com/LMMS/lmms/pull/6226#pullrequestreview-852504659

I think copy/paste might be broken too

**Comment by JohannesLorenz** [comment]
> Wait, so savefiles are already changed (and broken)? [#6226 (review)](https://github.com/LMMS/lmms/pull/6226#pullrequestreview-852504659)
> 
> I think copy/paste might be broken too

Sadly, yes, we messed this up. You probably need to wait for us to undo it until you can do it.

**Comment by allejok96** [comment]
@M374LX you will fix the broken project file loading while keeping the new names in savefiles, right? Then I'll just leave that TODO for you and look into copy/paste instead...

**Comment by allejok96** [comment]
I'm gonna indulge in BB -> Pattern, let's see how that goes

**Comment by M374LX** [comment]
#6279 addresses the breakage of data file upgrades.

@allejok96 The names in save files have not yet changed, but this is planned.

**Comment by allejok96** [comment]
Sorry @M374LX by "save files" do you mean the mmp[z] files created when you save? Because the XML tags *have* changed, but LMMS doesn't really care about the tag names in this case, so both old (not too old #6279) and new projects load just fine atm.

**Comment by M374LX** [comment]
@allejok96 I have now seen that some tag changes have slipped. These were unintentional and are being reverted through #6279, as the intention is not to change the date files at the moment.

**Comment by JohannesLorenz** [comment]
@allejok96 We merged a fix. If you want, you can write the `DataFile` patch now.

**Comment by allejok96** [comment]
Might as well do the whole DataFile thing in #6278 ... perhaps wait until after BB rename is merged and do both DataFile changes in one swoop?

**Comment by JohannesLorenz** [comment]
> Might as well do the whole DataFile thing in #6278 ... perhaps wait until after BB rename is merged and do both DataFile changes in one swoop?

:+1: for modular commits. Let's not wait here. After some tests I'll merge #6278 .

**Comment by PhysSong** [comment]
#6173(`enum` to `enum class`) looks like a good candidate to include for reorg.

**Comment by JohannesLorenz** [comment]
> #6173(`enum` to `enum class`) looks like a good candidate to include for reorg.

Rebasing over an `enum` to `enum class` should, in the worst case, require you adding some casts. I'm not sure if this must be bundled with reorg, which contains commits that are hard to rebase over.

**Comment by JohannesLorenz** [comment]
#6284 has been merged, and @allejok96 will now deliver the corresponding DataFile PR (which will also correspond to #6226 , it's the last "rename" PR). After that, we only have 3 things to do (IIRC, and considering the namespace as non-reorg):

1. Remove useless includes
2. Clang-Formatting
3. Move files

I recommend that we will start with (1), because it makes LMMS (hopefully) faster to compile, speeding up all the re-compiling for (2) and (3). I also volunteer to do (1), since it would be a lot of fun for me.

**Comment by Spekular** [comment]
We're also still missing a data file upgrade for the `tco` -> `clip` rename, aren't we?

**Comment by allejok96** [comment]
@Spekular #6309 does the `bb` and `tco` renaming in the same swoop

**Comment by JohannesLorenz** [comment]
Finally, all "rename PRs" done.

The next slot will be removing useless includes (issue: #5860). I'll try that out and submit a PR the next days.

**Comment by JohannesLorenz** [comment]
#5860 has been merged. The next (and probably the last difficult) 2 PRs are clang-tidy and clang-format. I'll try those.

**Comment by JohannesLorenz** [comment]
#6323 (clang-format) has been opened for review. It uses the conventions discussed/tested in #5948 .

**Comment by JohannesLorenz** [comment]
**Only "Reorganise all files into better places & folders" is left**. Not too difficult, but a few of those issues require some interpretation, like what should be considered a widget or not. Who can do that?

Note: Considering clang-format/clang-tidy, we agreed in discord that this will not be done in a reorg commit, but I will just provide the 2 config files and the devs can use them whenever they want.

**Comment by JohannesLorenz** [comment]
I'll do the 1st of those 4 points (renaming `LedCheckbox.h` to `LedCheckBox.h`, rename `stereo_enhancer` vs `stereoenhancer` etc).

**Comment by musikBear** [comment]
@JohannesLorenz The renaming job -isent that quite simple?
Just going through the files and doing those edits?
If there are no structural implications, then i volunteer to do some or all.
Just ping me

**Comment by JohannesLorenz** [comment]
> The renaming job -isent that quite simple?

Actually, I don't know. It's probably good if you have done something with these classes. Feel free to try points 2-4 if you understand these widgets.

**Comment by musikBear** [comment]
> Feel free to try points 2-4 if you understand these widgets.

Cant find these 3 and 4
You wrote
> I'll do the **1st of those 4 points** (renaming LedCheckbox.h to LedCheckBox.h, rename stereo_enhancer vs stereoenhancer etc).

So  LedCheckbox.h to LedCheckBox.h is **'1'**
And _stereoenhancer to stereo_enhancer_ is **'2'**
But what are **3 and 4** ?
I will start with **'2'**  -eg stereoenhancer to stereo_enhancer

**Comment by JohannesLorenz** [comment]
No, don't start. Read the top post: There are 4 points below "Reorganise all files into better places & folders". So I already have a PR for stereoEnhancer.

**Comment by musikBear** [comment]
There are manymany points in top-post
![image](https://user-images.githubusercontent.com/6662066/159337907-25897c58-5631-4842-8a49-abb51e626ecc.png)
and more
Please write in plain text what is is i have to rename in codebase

**Comment by allejok96** [comment]
When it comes to the `gui/widgets` folder we should make a list of all cpp files and decide what goes where

**Comment by JohannesLorenz** [comment]
The giant plugin rename PR #6350 has been merged.

The last part of re-org, the only remaining section, citing from the OP:

  - Reorganise all files into better places & folders (_requires some discussion_ #5944)
    - Move Mixer, Controller Rack, and Project Notes windows to the `gui/editors` folder
    - Move dialogs (that open in an external window) into `gui/dialog`. Delete the now-empty `gui/forms` folder
    - Remove widgets that can't be reused from `gui/widgets`, and move reusable widgets in. See [this comment]

This is all due to discussions in #5944 . @Spekular has created this bullet point and probably had some ideas for them, but in the last 2 weeks, no one has designed any plan.

Who wants to complete those last points (and is familiar with these classes)? Even a plan would be helpful. If no one is interested in the next 2 weeks, I'd say we'll skip these points completely.

**Comment by Spekular** [comment]
I'm willing to take on the editor and dialog moves as they're clearly defined, but my opportunities to work on LMMS are rather sporadic right now. I should have time tomorrow, so if I don't open a PR by the end of Monday, or if someone is able to open one today, I'll gladly relenquish that task.

I don't feel familiar enough with the widgets to complete that move in a reasonable timeframe. IMHO moving irrelevant classes out of the folder should be higher priority than moving relevant ones into it, if doing the whole thing is too big of a task.

**Comment by allejok96** [comment]
I'm in the process of compiling an overview of all widgets, as a base for discussion, and to better understand where they fit...

**Comment by ShalokShalom** [comment]
I like to know, why this is all done?

What's the point behind this reorganisation? 

Thanks a lot

**Comment by r4j4h** [comment]
@ShalokShalom [Based on the original post](https://github.com/LMMS/lmms/issues/120) this issue was started based on a goal for shorter, more univeral names. [There was a realization that a greater cleanup was needed, a la this post.](https://github.com/LMMS/lmms/issues/120#issuecomment-661714967)

**Comment by ShalokShalom** [comment]
So this is a simple renaming, and no improvement on the technical side?

**Comment by JohannesLorenz** [comment]
> So this is a simple renaming, and no improvement on the technical side?

It's not only renaming. The scope is explained in the top post.

Besides that, it's not useful to question a project which is ~95% completed...

**Comment by allejok96** [comment]
@ShalokShalom by renaming some things that people have requested for years, removing some inconsistencies and organizing the code in a better way, we're not only making the UI more approachable to new users, we're also making the code easier to understand and work with, and thus preventing future bugs.

If that doesn't count as a technical improvement, call it a technical investment.

**Comment by ShalokShalom** [comment]
Well, this is fine, and it takes such a year long effort to rename things?

**Comment by Veratil** [comment]
> Well, this is fine, and it takes such a year long effort to rename things? I also see some classes got reorganized, I still fail to see how renaming and reorganizing some classes is a "massive task" Is this really just because OO is so restrictive?

No one here works full or part time on LMMS, we're all volunteer time. Things get done when people have time to do it. Feel free to help out to get more things done if you think it's taking too long.

**Comment by ShalokShalom** [comment]
I was commenting on the amount of work to be done, you were answering on how many people do how much work. Off-topic.

**Comment by RoxasKH** [comment]
> I was commenting on the amount of work to be done, you were answering on how many people do how much work. Off-topic.

To be honest he was perfectly on-topic as what defines a task as "massive" is not only the amount of work but especially the time needed to get the work done, and what Veratil said describes exactly the context this situation is in, and that it's still a big change in the codebase considering the little number of devs, required reviews and all of that stuff.
On the contrary, your comment seems just negative (and surely not useful), at least under my eyes. I think anyone can get that anything that's called "reorganization" doesn't imply improvements on the technical side, in my opinion.
But that doesn't either mean it's not useful, or doesn't bring improvements in general.

Also i'm pretty sure that in most software engineering or software design and analysis courses, it's said that code reorganizations are usually the most demanding tasks, as they gets thing stopped and affect all the codebase, but the benefits that they give are essential to get a quicker job (other than mantaining a certain understanding of the code and code quality) in the future.

Saying it's offtopic seems just clutching at straws, as the things are stricly related; also, offtopic on what, on your question? The whole topic here is about the reorganization, and the fact the development isn't as active as everyone would like to, it's part of the whole development process, including this reorg. Even if it was just renaming that doesn't mean it's a little things, depends on how many things you have to rename.

In conclusion, as Veratil said, if you think this task doesn't seem massive for you and that you could get the job done quicker than how it is now, that's really nothing that is stopping you to help getting it done quicker.

**Comment by ShalokShalom** [comment]
So its not that inheritance, and the dependence of state within a tree is responsible for it, and that composition makes simple refactoring easier? 

Since thats what the literature says.

**Comment by sakertooth** [comment]
For the livelihood of LMMS, this is what I would personally like to see happen. I do not expect this to happen overnight, but with patience I think it can happen. 

A rewrite of the current implementation.

The interface can mostly stay the same (header organization for the most part), as if the interface changes then it will be a major headache trying to make the overall codebase tie back in.

However, some classes most certainly do need an interface change. There are many examples, but since I have recently been working on sample caching, I have come to realize that ``SampleBuffer`` is in need of a rework.

From my perspective, ``SampleBuffer`` is a class that needs **major** decoupling. There are functions that do not really apply to what the ``SampleBuffer`` should do. Why must a ``SampleBuffer`` be responsible for decoding and encoding Base64? Or why must it be responsible for opening audio files?

A ``SampleBuffer`` should be just that. A concrete class for temporary storage for samples which are in use by the current project. The functions in this class should pertain to this goal and that goal alone. If that happened to be the case, theres a greater chance for more maintable code.

The problem is that ``SampleBuffer``'s coupled interface is used throughout the codebase.  Changes in ``SampleBuffer`` will affect the classes that use it and its interface, which may affect the classes that use it and its interface, and so on and so fourth.

As such, the rewrite should be incremental. A bottom to top approach with small changes and moderate testing in between to make sure there is a gradual and not abrupt shift into the functioning of the application.

On the topic of implementation though, this is where I think most of the work should be in. Using ``SampleBuffer`` as an example once again, correct me if I am wrong, but I do not see the need for it to be a ``sharedObject`` any longer (let alone ``sharedObject`` existing in the codebase). ``std::shared_ptr`` and ``QSharedPointer`` exists, and it can be assigned a custom deleter as to which ``sharedObject`` seems to use ``object->deleteLater()``.

These are just my ideas. I would like to know what everybody else thinks about this if possible.

**Comment by M374LX** [comment]
Issues like #6287 prove that ``SampleBuffer`` needs a rework.

**Comment by JohannesLorenz** [comment]
> Issues like #6287 prove that `SampleBuffer` needs a rework.

Possibly, but does that rework need to be included in the re-org?

**Comment by M374LX** [comment]
The rework of `SampleBuffer` does not need to be necessarily included in the re-org.

**Comment by Rossmaxx** [comment]
#6374 has been merged. namespacing and we are ready to close it off.

**Comment by Rossmaxx** [comment]
Closing?

**Comment by Rossmaxx** [comment]
> As there have been many questions on why we have not closed it yet, I updated the list with the new, remaining tasks. Should all be doable...

Aren't they done, or are there anything else left?

**Comment by zonkmachine** [comment]
Ooopsie!

**Comment by issu-shooter** [comment]
Yes, the nightly version offers better performance and the GUI is somehow crisper, less blurry. I use the nightly myself, and gave people the 1.3.0 installer when they ask me for a free music studio software.

**Comment by allejok96** [comment]
> its about time to update 

Agreed. But back to the topic, what's the plan for the reorg now? I know there's many changes underway related to decoupling qt from core and other big low-level cleanup efforts that aren't part of this reorg really. Devs haven't been sleeping. There has probably been lot of discussion on discord, but I wanted to ask here.

So what's the plan? Should we consider the reorg over and see if something can be done with the stale PRs. Ping the authors? Or should we just forget about that bullet point and let authors update their PRs in due time if they still have any interest?

Is there anything left that I'm forgetting?

**Comment by issu-shooter** [comment]
Hmm, After allejok96 asked about what the plan is, it's been almost three months... Did the maintainers have forgotten this issue? Hope to see progress on this, everyone expects an update.

**Comment by sakertooth** [comment]
> Hmm, After allejok96 asked about what the plan is, it's been almost three months... Did the maintainers have forgotten this issue? Hope to see progress on this, everyone expects an update.

Not a maintainer, but it seems as if the reorg is essentially complete. The team was held back a bit by a couple of clang-tidy PRs, but #6564 should be the last one, as I haven't heard from @JohannesLorenz or @irrenhaus3 about adding anymore.

The plan after reorg is anybody's guess, but there are specific goals we have in mind, such as decoupling Qt from the core, as previously mentioned by allejok. I personally have ideas to start refactoring ``SampleBuffer`` little by little and simplify some of my PRs if possible.

The maintainers most likely did not forget about the reorg, but I agree that an update on the plan forward would be ideal.

**Comment by JohannesLorenz** [comment]
I just went through all checks from clang-tidy that I would like to add, and non of them will have reorg-scope. So I am OK with finishing the clang-tidy-checks (and reorg).

@allejok96 @DomClark @PhysSong @Veratil do you have any more wishes for clang-tidy-checks or re-org in general? Otherwise I would say we can finish it.

**Comment by allejok96** [comment]
If everyone's happy with the clang stuff I'd say lets move on. The main focus of our devs will probably still be related to refactoring and rewriting code, but IMHO this huge checklist can be closed.

I don't know what that will mean for the post re-org PRs, but I find it hard to believe features will come raining down once we officially end the re-org, although it would be a welcome change :smile:  As I see it the bottleneck of LMMS development hasn't been as much related to the re-org, as it's been to the lack of people with enough spare time and commitment to fix our big painful issues and - maybe even more important and boring - lack of people with the skills needed to review everything...

Boring-code devs, this is for you :trophy: 

*And for all you eager folks who's been sitting around and waiting for this to finish: don't get your expectations up too high. Instead read up on coding, and help us out in whatever way you can*

**Comment by DomClark** [comment]
I'm happy to move on. I have no objection to applying more clang-tidy fixes in the future, but don't feel the need to keep them as part of this one ongoing effort.

However, one thing I think would be worth doing is adding CI checks for the clang-tidy fixes that we have already applied. Otherwise there's no way to guarantee that we're still following those rules and we risk needing to do this all over again in the future.

**Comment by PhysSong** [comment]
I also agree with @DomClark's suggestion.

**Comment by JohannesLorenz** [comment]
I will write a PR for the suggested `.clang-tidy` update.

**Comment by JohannesLorenz** [comment]
> I will write a PR for the suggested `.clang-tidy` update.

#6619 . Would be nice if someone could review this one-file-PR, so we can finish re-org :grin:

**Comment by JohannesLorenz** [comment]
So, there is no more code to change for reorg :tada: We are currently discussing in Discord (dev-only) how to proceed with the "Reopen code for fresh PRs (or rebased old ones)" point.

**Comment by ryuukumar** [comment]
Following our discussion on Discord, this re-org is now complete, and this issue can be closed. Congrats to everyone who worked on this project! 🎉
